### ISSUE TITLE
Polygon smoothing generates artifacts since panda3D 1.10.x

### ISSUE THREAD

**Issue by el-dee** [panda3d/panda3d#993]
## Description

Since Panda3D 1.10.x using M_Polygon antialiasing mode generates artefacts.

See https://discourse.panda3d.org/t/holes-between-mesh-triangles-after-upgrading/24203/3 for the whole history and screenshots

As per the discussion, one possible outcome would be to disable the polygon smoothing if it is unreliable with modern HW.

## Steps to Reproduce

In Ralph sample, add the following line :

    render.setAntialias(AntialiasAttrib.M_polygon)

## Environment
 * Operating system: Linux
 * System architecture: amd64
 * Panda3D version: 1.10.x
 * Installation method: SDK
 * Python version (if using Python): Python 3.8
 * Compiler (if using C++): N/A

**Comment by el-dee** [comment]
No, the issue is that black lines appear at the limit of the triangles, looking like a weird frame mode.

Strangely I could not reproduce it anymore, even using old 1.10.x dev build, so I googled around and it is actually a bug/limitation of the NVidia driver. The algorithm used by NVidia to perform polygon smoothing is sensitive to the rendering order, the distance between the triangles, if they are reusing vertices or not, the depth test used, ... Also use of saturate blending is recommended (GL_SRC_ALPHA_SATURATE)

Here is the issue on another engine : https://gamedev.stackexchange.com/questions/109622/why-do-i-see-spaces-between-polygons-in-opengl-only-on-an-nvidia-720m

I would say the sudden appearance in 1.10.x is probably just a coincidence, maybe influenced by some change in Panda that increase the probability of the artefacts to appear.

**Comment by el-dee** [comment]
I updated the issue title to be more accurate. I believe this issue can be closed as is (or maybe a warning could ba added in the documentation).

The weird dots are another problem I guess :)

**Comment by rdb** [comment]
I managed to reproduce this issue by enabling transparency.  Could you edit the issue to include that bit?  Or are you getting it without transparency as well?

One workaround is to use `M_dual` mode.  This renders the opaque pixels separately, without color blending enabled, so opaque geometry won't suffer from this issue.

From the OpenGL documentation:
> If enabled, draw polygons with proper filtering. Otherwise, draw aliased polygons. For correct antialiased polygons, an alpha buffer is needed and the polygons must be sorted front to back. 

> Polygon antialiasing is optimized using blend function (`GL_SRC_ALPHA_SATURATE`, `GL_ONE`) with polygons sorted from nearest to farthest.

Transparency, however, requires sorting back-to-front.  So, I think this is just a limitation of polygon smoothing.

I can't really seem to get GL_SRC_ALPHA_SATURATE to work.  I just get a black window.

If this is not due to a regression in Panda, maybe we should just automatically disable polygon smoothing for transparent geometry in `M_auto` mode.  Or just deprecate the feature entirely.

**Comment by el-dee** [comment]
> I managed to reproduce this issue by enabling transparency. Could you edit the issue to include that bit? Or are you getting it without transparency as well?

OK, I will add that info in the issue description. Honestly, I don't remember if it was occurring on a transparent or plain surface, and up to now I haven't been able to reproduce it, with or without transparent surface. (I switched long ago to MSAA and never really used anymore)

> If this is not due to a regression in Panda, maybe we should just automatically disable polygon smoothing for transparent geometry in `M_auto` mode. 

Perhaps it's for the best to disable it in M_auto mode for transparent geometry as, per OpenGL, it shouldn't work anyway. On the other hand, if one enable it explicitly we should assume they know what they are doing.

> Or just deprecate the feature entirely.

I doesn't seem so broken after all, and it's probably useful on older or more limited hardware that does not support MSAA, I guess it's really sensitive over the configuration.
